crypto/merklekeystore/persistent_keystore.go

```diff
@@ -0,0 +1,177 @@
+		// Should we count the number of rows affected?
+		row := tx.QueryRow("SELECT Key FROM BlockProofKeys WHERE Round = ?", round)
+		err := row.Scan(&keyB)
+		if err != nil {
+			return fmt.Errorf("failed to select blockProof key for round %d : %w", round, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("PersistedParticipation.GetKey: %w", err)
+	}
+
+	key := &crypto.SignatureAlgorithm{}
+	err = protocol.Decode(keyB, key)
+	if err != nil {
+		return nil, fmt.Errorf("PersistedParticipation.GetKey: %w", err)
+	}
+
+	return key, nil
+}
+
+// DeleteKey deletes the specified key from the database and from cache if necessary
+func (p *PersistentKeystore) DeleteKey(round uint64) error {
+	return p.store.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		res, err := tx.Exec("DELETE FROM BlockProofKeys WHERE round = ?", round)
+		if err == nil {
+			count, err := res.RowsAffected()
+			if err == nil {
+				if count != 1 {
+					return fmt.Errorf("tried to delete a single key for round ?, instead deleted ? keys", round, count)
+				}
+			}
+		}
+		return err
+	})
+}
+
+func (p *PersistentKeystore) Length() (count int, err error) {
+	err = p.store.Atomic(func(ctx context.Context, tx *sql.Tx) error { // does it take too long?
+		row := tx.QueryRow("SELECT COUNT(*) FROM BlockProofKeys")
+		err = row.Scan(&count)
+		if err != nil {
+			return fmt.Errorf("failed to count rows in table BlockProofKeys : %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return -1, err
+	}
+
+	return count, nil
+}
+
+func keystoreInstallDatabase(tx *sql.Tx) error {
+	_, err := tx.Exec(`CREATE TABLE BlockProofKeys (
+    	id	  INTEGER PRIMARY KEY, 
+    	round INTEGER,	    --*  committed round for this key
+		blockProofKey BLOB  --*  msgpack encoding of ParticipationAccount.BlockProof.SignatureAlgorithm
+		);`)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO schema (tablename, version) VALUES (?, ?)", keystoreTableSchemaName, keystoreSchemaVersion)
+
+	return err
+}
+
+func RestoreKeystore(store db.Accessor) (PersistentKeystore, error) {
+	err := store.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return MigrateDB(tx)
+	})
+	if err != nil {
+		return PersistentKeystore{}, err
+	}
+
+	return PersistentKeystore{store}, nil
+}
+
+func MigrateDB(tx *sql.Tx) error {
+	var exists bool
+	schemaQuery := fmt.Sprintf("SELECT version FROM schema WHERE tablename = ?", keystoreTableSchemaName)
+	tx.QueryRow("SELECT EXISTS(%s)", schemaQuery).Scan(&exists)
+	if !exists {
+		// In the future this should not return quietly, as blockproof keys will be required
+		return nil
+	}
+
+	var version int
+	row := tx.QueryRow(schemaQuery)
+	err := row.Scan(&version)
+	if err != nil {
+		// fmt.Errorf("restore (agreement): could not query raw state: %v", err)
+		return err
+	}
+
+	err = updateDB(tx, version)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updateDB(tx *sql.Tx, version int) error {
```

Comment:
remove this

Reply:
This is where the actual migrations will be implemented, it is currently empty since there are no versions for this table.
This pattern is similar to the one in `partInstall.go`